torch_geometric/transforms/add_metapaths.py

```diff
@@ -145,3 +155,20 @@ def __call__(self, data: HeteroData) -> HeteroData:
                     del data[node]
 
         return data
+
+    def _edge_index_sampling(self, edge_index, max_sample, backward=False):
+        if backward:
+            node_idx = edge_index[1]
+        else:
+            node_idx = edge_index[0]
+        _, counts = torch.unique(node_idx, sorted=False, return_counts=True)
+        sample_prob = torch.repeat_interleave(1 / counts * max_sample, counts)
+        draw = torch.rand(sample_prob.size(), dtype=float)
+        mask = sample_prob > draw
+        return edge_index[:, mask]
+
+    def edge_index_sampling(self, edge_index, max_sample):
+        edge_index = self._edge_index_sampling(edge_index, max_sample)
+        # sample the reverse direction: restrict the destination degree
+        # to max_sample in expectation
+        return self._edge_index_sampling(edge_index, max_sample, backward=True)
```

Comment:
Since it's a bipartite edge_index, I sampled twice. The first pass ensures that source degree is upper-bounded; the second pass ensures that dst degree is upper-bounded.
It will be the same as limiting both the adj row and column sums instead of just rows.